processor/lsmintervalprocessor/internal/merger/limits/tracker.go

```diff
@@ -0,0 +1,297 @@
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package limits // import "github.com/elastic/opentelemetry-collector-components/processor/lsmintervalprocessor/internal/merger/limits"
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"hash"
+	"slices"
+
+	"github.com/axiomhq/hyperloglog"
+)
+
+// Tracker tracks the configured limits while merging. It records the
+// observed count as well as the unique overflow counts.
+type Tracker struct {
+	maxCardinality uint64
+	// Note that overflow buckets will NOT be counted in observed count
+	// though, overflow buckets can have overflow of their own.
+	observedCount  uint64
+	overflowCounts *hyperloglog.Sketch
+}
+
+func newTracker(maxCardinality uint64) *Tracker {
+	return &Tracker{maxCardinality: maxCardinality}
+}
+
+func (t *Tracker) Equal(other *Tracker) bool {
+	if t.maxCardinality != other.maxCardinality {
+		return false
+	}
+	if t.observedCount != other.observedCount {
+		return false
+	}
+	return t.EstimateOverflow() == other.EstimateOverflow()
+}
+
+func (t *Tracker) HasOverflow() bool {
+	return t.overflowCounts != nil
+}
+
+func (t *Tracker) EstimateOverflow() uint64 {
+	if t.overflowCounts == nil {
+		return 0
+	}
+	return t.overflowCounts.Estimate()
+}
+
+// CheckOverflow checks if overflow will happen on addition of a new entry with
+// the provided hash denoting the entries ID. It assumes that any entry passed
+// to this method is a NEW entry and the check for this is left to the caller.
+func (t *Tracker) CheckOverflow(f func() hash.Hash64) bool {
+	if t.maxCardinality == 0 {
+		return false
+	}
+	if t.observedCount == t.maxCardinality {
+		if t.overflowCounts == nil {
+			// Creates an overflow with 14 precision
+			t.overflowCounts = hyperloglog.New14()
+		}
+		t.overflowCounts.InsertHash(f().Sum64())
+		return true
+	}
+	t.observedCount++
+	return false
+}
+
+// MergeEstimators merges the overflow estimators for the two trackers.
+// Note that other required maintenance of the tracker for merge needs to
+// done by the caller.
+func (t *Tracker) MergeEstimators(other *Tracker) error {
+	if other.overflowCounts == nil {
+		// nothing to merge
+		return nil
+	}
+	if t.overflowCounts == nil {
+		t.overflowCounts = other.overflowCounts.Clone()
+		return nil
+	}
+	return t.overflowCounts.Merge(other.overflowCounts)
+}
+
+// Marshal marshals the tracker to a byte slice.
+func (t *Tracker) Marshal() ([]byte, error) {
```

Comment:
Should this also be `AppendBinary`? Then we could avoid a slice allocation & copy in `Value.Marshal`

Reply:
Sounds good. We won't be able to fully utilize this though since it is not possible to estimate the size of the tracker prior to marshaling, so we could still end up reallocating the slice. Also see: https://github.com/axiomhq/hyperloglog/issues/44

Reply:
Yeah no worries, we can follow up on this